### PR TITLE
[REVIEW] Create tables from other RDBMS

### PR DESCRIPTION
**Pull request by aucahuasi**
https://github.com/BlazingDB/blazingsql/issues/1395

### CODE REVIEW

**Comment by roaramburu** [comment]
What does everyone think of the argument `from_sql`?

@drabastomek what do you think? I think `engine` might be better and more informative.

cc @felipeblazing @williamBlazing

**Comment by drabastomek** [comment]
@aucahuasi @roaramburu 

I think `from_sql` might be a bit misleading as someone might think this is ok:

```python
bc.from_sql(bc.sql('SELECT * FROM <foo>')
```

How about `from_db` or `from_db_table`?

File: engine/src/cython/engine.cpp [hunk 1]
```diff
@@ -65,22 +77,39 @@ std::pair<std::vector<ral::io::data_loader>, std::vector<ral::io::Schema>> get_l
...
 		} else if(fileType == ral::io::DataType::ARROW){
 	     	parser = std::make_shared<ral::io::arrow_parser>(tableSchema.arrow_table);
-		}
+		} else if(fileType == ral::io::DataType::MYSQL) {
+#ifdef MYSQL_SUPPORT
+      parser = std::make_shared<ral::io::mysql_parser>();
+      auto sql = ral::io::getSqlInfo(args_map);
+      provider = std::make_shared<ral::io::mysql_data_provider>(sql);
+      isSqlProvider = true;
+#endif
```
Comment (williamBlazing) [inline engine/src/cython/engine.cpp:86]: there should be an else to the `#ifdef MYSQL_SUPPORT` so that if its not compiled with that support and someone tries to make a table with that, that is outputs an appropriate error

File: engine/src/cython/engine.cpp [hunk 2]
```diff
@@ -65,22 +77,39 @@ std::pair<std::vector<ral::io::data_loader>, std::vector<ral::io::Schema>> get_l
...
+    } else if(fileType == ral::io::DataType::SQLITE) {
+#ifdef SQLITE_SUPPORT
+  //		parser = std::make_shared<ral::io::sqlite_parser>();
+  //    auto sql = ral::io::getSqlInfo(args_map);
+  //    provider = std::make_shared<ral::io::sqlite_data_provider>(sql.schema,
+  //                                                               sql.table,
+  //                                                               sql.table_filter,                                                          sql.table_batch_size);
+  //    isSqlProvider = true;
+#endif
+    }
```
Comment (williamBlazing) [inline engine/src/cython/engine.cpp:96]: why is postgressql not in this list?

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/cython/engine.cpp (hunk 2 above)
Comment (felipeblazing) [reply engine/src/cython/engine.cpp:96]: i think its still in progress for a seperate pr

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp [hunk 3]
```diff
@@ -123,6 +124,11 @@ kstatus TableScan::run() {
     std::vector<int> projections(schema.get_num_columns());
     std::iota(projections.begin(), projections.end(), 0);
 
+    if (provider->is_sql()) {
+      auto sql_provider = std::dynamic_pointer_cast<ral::io::abstractsql_data_provider>(provider);
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:128]: This should happen at construction since we can determine that the moment the provider is created so it is not needed to do this casting or set these column indices here. that should happen during construction

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp [hunk 4]
```diff
@@ -266,6 +272,11 @@ kstatus BindableTableScan::run() {
         std::iota(projections.begin(), projections.end(), 0);
     }
 
+    if (provider->is_sql()) {
+      auto sql_provider = std::dynamic_pointer_cast<ral::io::abstractsql_data_provider>(provider);
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:276]: same as comment right above, this should be set at construction

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp [hunk 5]
```diff
@@ -278,6 +289,11 @@ kstatus BindableTableScan::run() {
             //this will allow us to prevent from having too many open file handles by being
             //able to limit the number of file tasks
             auto handle = provider->get_next(true);
+            if (provider->is_sql()) {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:292]: you should not be adding any special logic here. This part of the code needs to be completely ignorant of what kind of provider it is

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 3 above)
Comment (felipeblazing) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:128]: also and more importantly, the kernels shoudl have no idea of what kind of provider they have or dont have. They should only use the functions that are common to the provider base class. That is part of the intention of these classes. Yoiu shouldnt need to add anything to them

File: engine/src/io/data_parser/ArgsUtil.cpp [hunk 6]
```diff
@@ -242,11 +249,53 @@ std::string getDataTypeName(DataType dataType) {
...
+	case DataType::SQLITE: return "sqlite"; break;
 	default: break;
 	}
 
 	return "undefined";
 }
 
+sql_info getSqlInfo(std::map<std::string, std::string> &args_map) {
+  // TODO percy william maybe we can move this constant as a bc.BlazingContext config opt
+  const size_t DETAULT_TABLE_BATCH_SIZE = 1000;
```
Comment (felipeblazing) [inline engine/src/io/data_parser/ArgsUtil.cpp:263]: That feels like a very small number depending on the latency and overhead of fetching lets say 10k or 100k rows vs fetching 1k 10 or 100 times

File: engine/src/cython/io.cpp [hunk 7]
```diff
@@ -47,41 +66,65 @@ TableSchema parseSchema(std::vector<std::string> files,
...
-			}
+    if (isSqlProvider) {
+        ral::io::data_handle handle = provider->get_next(false);
+        parser->parse_schema(handle, schema);
+        if (schema.get_num_columns() > 0){
+          got_schema = true;
+        }
+    } else {
+      while (!got_schema && provider->has_next()){
+        ral::io::data_handle handle = provider->get_next();
```
Comment (williamBlazing) [inline engine/src/cython/io.cpp:105]: What if we make `data_handle` an interface instead of a struct where we are shoving more mutually exclusive stuff into it?

If we did that, we could replace `handle.file_handle != nullptr` with `handle.is_valid()` which would apply to any type of data_handle and make a lot of other things cleaner

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp [hunk 8]
```diff
@@ -123,6 +124,11 @@ kstatus TableScan::run() {
     std::vector<int> projections(schema.get_num_columns());
     std::iota(projections.begin(), projections.end(), 0);
 
+    if (provider->is_sql()) {
+      auto sql_provider = std::dynamic_pointer_cast<ral::io::abstractsql_data_provider>(provider);
+      sql_provider->set_column_indices(projections);
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:129]: this logic of setting the column indices that you want should be handled by the CacheDataIO. all the other data types handle it there.

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 4 above)
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:276]: this logic of setting the column indices that you want should be handled by the CacheDataIO. all the other data types handle it there.

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp [hunk 9]
```diff
@@ -278,6 +289,11 @@ kstatus BindableTableScan::run() {
             //this will allow us to prevent from having too many open file handles by being
             //able to limit the number of file tasks
             auto handle = provider->get_next(true);
+            if (provider->is_sql()) {
+              if (handle.sql_handle.row_count == 0) {
```
Comment (williamBlazing) [inline engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:293]: This should be handled by the `has_next()` function that was just called. If it returned true, then the row count here should never be 0. Is there a reason we are doing this?

File: engine/src/io/data_parser/sql/AbstractSQLParser.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "io/data_parser/DataParser.h"
+#include "arrow/io/interfaces.h"
+#include <memory>
+#include <vector>
+
+namespace ral {
+namespace io {
+
+class abstractsql_parser : public data_parser {
```
Comment (williamBlazing) [inline engine/src/io/data_parser/sql/AbstractSQLParser.h:17]: What is the purpose of this class? It looks almost the same as the original data_parser interface?

Comment (williamBlazing) [reply engine/src/io/data_parser/sql/AbstractSQLParser.h:17]: for that matter, it looks like its not used anywhere. Please delete it if its not being used

File: engine/src/io/data_parser/sql/MySQLParser.cpp [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+        std::vector<int32_t> *tmp = new std::vector<int32_t>();
+        tmp->resize(total_rows);
+        host_cols[col] = tmp;
+      } break;
+      case cudf::type_id::INT64: {
+        std::vector<int64_t> *tmp = new std::vector<int64_t>();
+        tmp->resize(total_rows);
+        host_cols[col] = tmp;
+      } break;
+      case cudf::type_id::UINT8: {
```
Comment (williamBlazing) [inline engine/src/io/data_parser/sql/MySQLParser.cpp:131]: Hmmmm, have you tried Unsigned integers? BSQL does not support unsigned integers yet. Not sure what would happen if you tried using them here.  I would guess we should leave this part here as is, and in the schema parsing, make sure that it maps to a signed integer

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 11 above)
Comment (felipeblazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:131]: what we did before was make the smller types larger. though now that i mention it this is just the allocation logic basically. What is the purpose of allocating a std vector then casting it to void? Wouldnt it be the same to just allocate the space?

File: engine/src/io/data_parser/sql/MySQLParser.cpp [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+      case cudf::type_id::LIST: {} break;
+      case cudf::type_id::DECIMAL32: {} break;
+      case cudf::type_id::DECIMAL64: {} break;
+      case cudf::type_id::STRUCT: {} break;
+      case cudf::type_id::NUM_TYPE_IDS: {} break;
+    }
+  }
+
+  int row = 0;
+  while (res->next()) {
```
Comment (felipeblazing) [inline engine/src/io/data_parser/sql/MySQLParser.cpp:193]: can mysql c api only get one row at a time?

File: engine/src/io/data_parser/sql/MySQLParser.cpp [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+        case cudf::type_id::TIMESTAMP_NANOSECONDS:
+        // TODO percy it seems we don't support this yet
+        // case cudf::type_id::DURATION_DAYS: {} break;
+        // case cudf::type_id::DURATION_SECONDS: {} break;
+        // case cudf::type_id::DURATION_MILLISECONDS: {} break;
+        // case cudf::type_id::DURATION_MICROSECONDS: {} break;
+        // case cudf::type_id::DURATION_NANOSECONDS: {} break;
+        // case cudf::type_id::DICTIONARY32: {} break;
+        case cudf::type_id::STRING: {
+          std::string real_data = res->getString(mysql_col).asStdString();
```
Comment (felipeblazing) [inline engine/src/io/data_parser/sql/MySQLParser.cpp:320]: it might be faster to accumulate a vector of strings and make a cudf string column once you have a batch full

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+        if (fsize < 0) {  // STRING, STRUCT, LIST, and similar cases
+          auto *string_col = new cudf_string_col();
+          string_col->offsets.reserve(resultNtuples + 1);
+          string_col->offsets.push_back(0);
+          return static_cast<void *>(string_col);
+        }
+        // primitives cases
+        const cudf::type_id cudf_type_id = cudf_types[projection_index];
+        switch (cudf_type_id) {
+        case cudf::type_id::INT8: {
```
Comment (felipeblazing) [inline engine/src/io/data_parser/sql/PostgreSQLParser.cpp:71]: alot of this code seems to be shared with the mysqlparser perhaps the abstract_sql_parser should implement some methods :)

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp [hunk 15]
```diff
@@ -0,0 +1,549 @@
...
+  }
+
+  cudf::io::table_with_metadata tableWithMetadata;
+  std::vector<std::unique_ptr<cudf::column>> cudf_columns;
+  cudf_columns.resize(static_cast<std::size_t>(resultNfields));
+  for (const std::size_t projection_index : column_indices) {
+    cudf::type_id cudf_type_id = cudf_types[projection_index];
+    switch (cudf_type_id) {
+    case cudf::type_id::INT8: {
+      std::vector<std::int8_t> *vector =
```
Comment (felipeblazing) [inline engine/src/io/data_parser/sql/PostgreSQLParser.cpp:286]: this whole pattern with the void * feels fragile i wonder if just managing the memory and types is better

File: engine/src/io/data_parser/sql/sqlcommon.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+                   const std::vector<cudf::bitmask_type> &null_mask) {
+  auto d_chars = cudf::detail::make_device_uvector_sync(host_col->chars);
+  auto d_offsets = cudf::detail::make_device_uvector_sync(host_col->offsets);
+  auto d_bitmask = cudf::detail::make_device_uvector_sync(null_mask);
+  return cudf::make_strings_column(d_chars, d_offsets, d_bitmask);
+}
+
+template <typename T>
+static inline std::unique_ptr<cudf::column>
+build_fixed_width_cudf_col(size_t total_rows,
```
Comment (felipeblazing) [inline engine/src/io/data_parser/sql/sqlcommon.h:27]: i think there are already cudf helpers for this and this doesnt seem to be related to sql

File: engine/src/cython/io.cpp [hunk 17]
```diff
@@ -47,41 +66,65 @@ TableSchema parseSchema(std::vector<std::string> files,
...
 		bool open_file = false;
-		while (provider->has_next()){
-			std::vector<ral::io::data_handle> handles = provider->get_some(64, open_file);
-			for(auto handle : handles) {
-				schema.add_file(handle.uri.toString(true));
-			}
-		}
+    if (!isSqlProvider) {
+      while (provider->has_next()){
+        std::vector<ral::io::data_handle> handles = provider->get_some(64, open_file);
```
Comment (felipeblazing) [inline engine/src/cython/io.cpp:122]: @williamBlazing it seems to be used here the abstractsql_dat_provider provides this method so its not just decorative

**Review by felipeblazing — CHANGES_REQUESTED** [review]

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 11 above)
Comment (williamBlazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:131]: I would recommend making `host_cols` a `std::vector<std::vector<char>>` so that you can just set the size of each `std::vector<char>` the size in bytes that it needs to hold the data and you don't need to do `new` which can lead to memory leaks if not handled correctly. You can then do a reinterpret_cast when setting or getting the values.

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 8 above)
Comment (williamBlazing) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:129]: @felipeblazing what do you think?

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 9 above)
Comment (williamBlazing) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp:293]: Ok, I looked at the code and understand why you need to do this. What I would suggest is that `mysql_data_provider` be able to always hold a result set. For example:
It could have two extra members:
```
data_handle current_handle;
bool have_current_handle;  // initialized to false
```
then
```
bool mysql_data_provider::has_next() {
    if (!this->table_fetch_completed  && !have_current_handle){
         fetch_data();
   }
  return (this->table_fetch_completed == false);
}
```
`fetch_data()` would have the logic that is in `get_next()` that obtains and builds the data_handle.
`have_current_handle` would be set to true by `fetch_data` and set to false by `get_next`.

You can then take it to the next step and make it so that when you call `get_next()` after it returns the handle, it runs `fetch_data` on a separate thread so that it is already fetching the next batch while the rest of the engine does parsing or anything else. This async behaviour could be configurable

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 12 above)
Comment (williamBlazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:193]: With regards to the range checking, if we have a value outside the range, we would just have to throw an error. We cant change the type midway. But that is better than nothing. I think we will implement Unsigned types soon, so this will hopefully be a short term issue

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 14 above)
Comment (williamBlazing) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:71]: i agree

File: pyblazing/pyblazing/apiv2/context.py [hunk 18]
```diff
@@ -3053,6 +3094,27 @@ def sql(
                         query_table.input = [query_table.input]
                     currentTableNodes.append(query_table)
 
+            elif (
```
Comment (williamBlazing) [inline pyblazing/pyblazing/apiv2/context.py:3097]: what is actually happening here? this is just copied from when you have a file. How does the engine distribute pulling data from mySQL from different nodes?

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 12 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:193]: @felipeblazing yes unfortunately mysql is using the jdbc approach and there is nothing like fetch_bacth in the jdbc standard, I didn't find anything like "fetch_batch" in the native C neither. The good news is that I think the underlying implementation of those drivers have a buffer of results and the client API is just returning the latest item that was fetched.

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 14 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:71]: yep that was the reason of those abstract classes (common spaces), but I din't have enough time to move the code :p, I'll be move it there cc @gcca

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 15 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:286]: totally agree, we already did that, our first approach was using typed vectors, but the casting was an issue, also the handling for strings is different.
Let me try again and see if we can get ride of void*

File: engine/src/io/data_parser/sql/sqlcommon.h (hunk 16 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/sqlcommon.h:27]: we could not found the builder or a sample code, perhaps we can move this to CudfUtils somewhere ... do you know where I can found this cudf helpers (there must not be the ones in cudf::testing) cc @rommelDB

File: io/src/FileSystem/Uri.h [hunk 19]
```diff
@@ -24,6 +24,11 @@ class Uri {
 		const std::string & authority,
 		const Path & path,
 		bool strict = false);  // convenience ctor, will use Uri(scheme, auth, Path, strict)
+  Uri(const std::string &scheme,
```
[the remark this answers is not in the document]
Comment (aucahuasi) [reply io/src/FileSystem/Uri.h:27]: To save a URI for the internal representation of the sql batch we fetched, something like
"mysql://db/table/part?columns" ... perhaps is not a good idea yet but is always a good thing to have URI for the resources we are handling (data_hanle.uri)

File: pyblazing/pyblazing/apiv2/context.py [hunk 20]
```diff
@@ -912,6 +917,15 @@ def kwargs_validation(kwargs, bc_api_str):
...
             "local_files",
             "get_metadata",
+            # SQL Engines arguments
+            "from_sql",
+            "hostname",
+            "port",
+            "username",
+            "password",
+            "schema",
+            "table_filter",
```
[the remark this answers is not in the document]
Comment (aucahuasi) [reply pyblazing/pyblazing/apiv2/context.py:927]: Even when the predicate push down is ready I would vote to leave this "pre-filter" feature there, because it offers more granular ways to create a sql table. cc @drabastomek @roaramburu @felipeblazing

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 11 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:131]: @williamBlazing that was the first implementation of `host_cols` and if we do that we need to iterate all the rows again to construct the buffer for the cudf column. The way is implemented right now is pretty good because we save like 2 loops over all the rows and the data structs are exactly what the cudf ctors want cc @felipeblazing @gcca

Comment (williamBlazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:131]: Perhaps I am not explaining myself. Right now you are holding the data in vectors of specific types that you are creating with `new`, but then you are type-erasing them and holding them as `void*` pointers in `host_cols`. My suggestion was that you replace the individual typed vectors you are creating with `new` and only use type erased vectors of type char that you dont have to use `new`. You can do a reinterpret_cast on those vectors of char to give them a type when you want to put a value or create a cudf from them. The purpose of this suggestion is so that you dont have to use `new` which can lead to memory leaks if not handled correctly

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 14 above)
Comment (williamBlazing) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:71]: I think this refactoring can be done when the postgresql or sqlite implementations are done

File: io/src/FileSystem/Uri.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (williamBlazing) [reply io/src/FileSystem/Uri.h:27]: ok

File: engine/src/cython/engine.cpp (hunk 1 above)
Comment (aucahuasi) [reply engine/src/cython/engine.cpp:88]: Done

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 3 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: I've removed is_sql from the data_provider interface, done

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 4 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: done

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 5 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: done

File: engine/src/io/data_parser/ArgsUtil.cpp (hunk 6 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/ArgsUtil.cpp]: now is 100k, done

File: engine/src/cython/engine.cpp [hunk 21]
```diff
@@ -65,22 +77,39 @@ std::pair<std::vector<ral::io::data_loader>, std::vector<ral::io::Schema>> get_l
...
 		}
 
-		if(fileType == ral::io::DataType::CUDF || fileType == ral::io::DataType::DASK_CUDF) {
-			// is gdf
-			provider = std::make_shared<ral::io::gdf_data_provider>(tableSchema.blazingTableViews, uri_values[i]);
-		} else {
-			// is file (this includes the case where fileType is UNDEFINED too)
-			provider = std::make_shared<ral::io::uri_data_provider>(uris, uri_values[i]);
-		}
+    if (!isSqlProvider) {
```
[the remark this answers is not in the document]
Comment (aucahuasi) [reply engine/src/cython/engine.cpp:106]: Is just a flag and it simplifies the if statements later, if we rely on fileType we will then need to do if fileType!=mysql && fileType!=postgre ...etc

File: engine/src/cython/io.cpp (hunk 7 above)
Comment (aucahuasi) [reply engine/src/cython/io.cpp:105]: Yes the idea is to have a nice interface here, however I think that is not part of the scope of this PR. There are many ways create a better `data_handle` concept and to really be data_source agnostic, but we lack the concept of data_source per se in our engine. I would propose to take this issue later, not for this PR cc @felipeblazing @gcca

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 8 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: Done, the logic is now in the kernel ctors and the kernel::run is agnostic to the data provider cc @felipeblazing @gcca

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 4 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: Done, now CacheDataIO handle that part

File: engine/src/execution_graph/logic_controllers/BatchProcessing.cpp (hunk 9 above)
Comment (aucahuasi) [reply engine/src/execution_graph/logic_controllers/BatchProcessing.cpp]: @williamBlazing It seems the `async` optimization would be a nice to have, I've already resolved the main issue regarding to the `is_sql`, the use of `next` and the use of `CacheDataIO`. How about we implement async data_providers in a next PR? cc @felipeblazing @gcca

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 13 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp:320]: I think that is what we are doing, we create and accumulate in host and the once is ready we create the cudf columns, perhaps we need more info about what do you have in mind here cc @gcca

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 15 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:285]: @felipeblazing @williamBlazing If we get ride of bool we would need to iterate all the rows later to have the references for cudf, I think `void*` is not ideal but is what it offers the best performance for now (i.e. it minimizes a lot the travels over all the rows, at least 2 times: one for getting the refs and the other for adapting the refs for the cudf column ctors) cc @gcca

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 11 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp]: So the issue is about code style I guess, the `new` is being used but we are also using the `delete` to avoid any leaks.
The approach you are suggesting was the first implementation and it didn't work, because some types cannot be casted even using `reinterpred_cast`, this will fails for long double and strings.
If the core issue is about the use of `new` then note that we just need to make sure we are deleting correctly, and that is the case here. cc @felipeblazing @gcca

File: engine/src/io/data_parser/sql/AbstractSQLParser.h (hunk 10 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/AbstractSQLParser.h:20]: `abstractsql_parser` is a _thing_ now, it allow us to define some common flows for all the sql parsers

Comment (aucahuasi) [reply engine/src/io/data_parser/sql/AbstractSQLParser.h:20]: done

File: engine/src/io/data_parser/sql/PostgreSQLParser.cpp (hunk 14 above)
Comment (aucahuasi) [reply engine/src/io/data_parser/sql/PostgreSQLParser.cpp:70]: Great idea!

File: engine/src/io/data_provider/sql/AbstractSQLDataProvider.h [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class abstractsql_data_provider : public data_provider {
+public:
+  abstractsql_data_provider(const sql_info &sql);
+
+  virtual ~abstractsql_data_provider();
+
+  /**
+	 * Tries to get up to num_files data_handles. We use this instead of a get_all() because if there are too many files, 
+	 * trying to get too many file handles will cause a crash. Using get_some() forces breaking up the process of getting file_handles.
```
[the remark this answers is not in the document]
Comment (aucahuasi) [reply engine/src/io/data_provider/sql/AbstractSQLDataProvider.h:39]: done

File: pyblazing/pyblazing/apiv2/context.py (hunk 18 above)
Comment (aucahuasi) [reply pyblazing/pyblazing/apiv2/context.py:3135]: This is the only remain issue, we keep working on this one cc @gcca

Comment (aucahuasi) [reply pyblazing/pyblazing/apiv2/context.py:3140]: ok I just sent a push, that should fix this issue ... tomorrow morning we are going to run some test to make sure distribution is fine, good catch @williamBlazing thanks!

File: engine/src/io/data_parser/sql/MySQLParser.cpp (hunk 11 above)
Comment (felipeblazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp]: hes suggestiong you use a std::vector<char> no new and delete needed but it has another issue. the <char> there has no guaranteed alignment so however you allocate that vector you need to make sure its aligned to the largest type you want to represent.

Comment (felipeblazing) [reply engine/src/io/data_parser/sql/MySQLParser.cpp]: I agree that the whole type erasing a vector of something only to reinterpret cast it later is more problematic than the need to manage the memory resource.. it should be at least 8 byte aligned on a 64 bit system if im not mistaken so until we have 128 bit integers it might not matter

Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp]: great thanks, yeah and we are taking care to manage the memory resource there in the abstract_sql_parser, soon in the next PR postgresql and sqlite will use that same code so there will be only a single place where we are using new and delete.

Comment (aucahuasi) [reply engine/src/io/data_parser/sql/MySQLParser.cpp]: One solution we can try in the future is to use std::shared_ptr<void> instead of a naked void*, we can use std::unique_ptr here but that would require mor code to create custom dtors (std::shared_ptr infers the type for us) 
However I think dealing with a better style/approach for this code will be subject for another PR, the way it is now do the job and doesn't leak

File: pyblazing/pyblazing/apiv2/context.py (hunk 18 above)
Comment (aucahuasi) [reply pyblazing/pyblazing/apiv2/context.py:3137]: Distribution is working fine now, thanks!

File: pyblazing/pyblazing/apiv2/context.py [hunk 23]
```diff
@@ -3090,6 +3134,27 @@ def sql(
                         query_table.input = [query_table.input]
                     currentTableNodes.append(query_table)
 
+            elif (
+                query_table.fileType == DataType.MYSQL
+                or query_table.fileType == DataType.SQLITE
+                # or query_table.fileType == DataType.
+            ):
+                if query_table.has_metadata():
```
Comment (williamBlazing) [inline pyblazing/pyblazing/apiv2/context.py:3142]: This block of code was just copied form elsewhere. It does work, but its misleading and confusing. I would recommend replacing it with the following, which would be more explicit about what its doing:

```
currentTableNodes = []
for i in range(0, numSlices):
                bt = BlazingTable(self.name, self.input, self.fileType, args=self.args)
                bt.column_names = self.column_names
                bt.column_types = self.column_types
                currentTableNodes.append(bt)
```

**Review by williamBlazing — CHANGES_REQUESTED** [review]

File: engine/src/cython/io.cpp [hunk 24]
```diff
@@ -47,41 +66,63 @@ TableSchema parseSchema(std::vector<std::string> files,
 		parser = std::make_shared<ral::io::json_parser>(args_map);
 	} else if(fileType == ral::io::DataType::CSV) {
 		parser = std::make_shared<ral::io::csv_parser>(args_map);
-	}
+	} else if(fileType == ral::io::DataType::MYSQL) {
+#ifdef MYSQL_SUPPORT
+		parser = std::make_shared<ral::io::mysql_parser>();
+    auto sql = ral::io::getSqlInfo(args_map);
+    provider = std::make_shared<ral::io::mysql_data_provider>(sql, 0, 0);
+#endif
```
[the remark this answers is not in the document]
Comment (aucahuasi) [reply engine/src/cython/io.cpp:74]: For now PR only supports MySQL, perhaps we can let @gcca to finish the upcoming PR and there we can have the kind of changes you are suggesting cc @felipeblazing

File: pyblazing/pyblazing/apiv2/context.py (hunk 23 above)
Comment (aucahuasi) [reply pyblazing/pyblazing/apiv2/context.py:3142]: I would apply this changes in the up coming refactor project for pyblazing, as you said the code works but the style can improve, how about we push this suggestion for the next PRs cc @drabastomek

**Review by felipeblazing — APPROVED** [review]
